### PR TITLE
chore: Pinned workspaces in stores/workspaces [WEB-586]

### PR DESCRIPTION
**Pull request by mapmeld**
## Description

Incorporates filters into `useWorkspaces` so that it can take the place of `useFetchPinnedWorkspaces`. This includes shifting the internal state from `workspaces` (the `Loadable<Workspace[]>`) to a workspacesIndex `Map<string, Loadable<Workspace[]>>`

Work in progress while addressing the experiment store PR (confirming this use of Immutable Map), and with Move Experiment modal.

## Test Plan

- Pin at least one workspace
- Load Determined and confirm that the pinned workspace is shown on the left navigation
- When you refresh the page, the left navigation briefly shows a circle spinner before loading pinned workspaces
- Unpin the workspace and confirm that it disappears from the left navigation within a few seconds
- On the Workspaces view, each workspace has a card. Use the menu on the card to pin and unpin a workspace. The menu label should show 'pin workspace' and 'unpin workspace'
- Open an experiment. Select another workspace and project to move it.

## Checklist
- [x] Changes have been manually QA'd
- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [x] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: webui/react/src/components/NavigationSideBar.tsx [hunk 1]
```diff
@@ -180,6 +181,8 @@ const NavigationSideBar: React.FC = () => {
     openWorkspaceCreateModal();
   }, [openWorkspaceCreateModal]);
 
+  const pinnedWorkspaces = Loadable.getOrElse([], useWorkspaces({ limit: 0, pinned: true }));
```
Comment (ClaireNeveu) [inline webui/react/src/components/NavigationSideBar.tsx:184]: If you're not going to handle the loading states in this PR can you leave TODOs where they need to be handled?

**Review by ClaireNeveu — CHANGES_REQUESTED** [review]
> Unpin the workspace and confirm that it disappears from the left navigation within a few seconds

I know this isn't because of this PR but this would be a good place to fix this. We shouldn't be relying on polling to update the UI state when an action is taken by the user.

We can add a new function `update: (id: number, fn: Workspace => Workspace): void` to the workspace store that allows you to update a workspace in the store and then use that to update the state when pinning. We can then delete the polling for pinned workspaces.

You may need to change the underlying state to be
```typescript
workspaces: Map<number, Workspace>
workspacesByParams: Map<string, number[]>
```
to make that work.

File: webui/react/src/stores/workspaces.tsx [hunk 2]
```diff
@@ -53,15 +60,40 @@ export const useEnsureWorkspacesFetched = (canceler: AbortController): (() => Pr
...
 };
 
-export const useWorkspaces = (): Loadable<Workspace[]> => {
+export const useWorkspaces = (params?: GetWorkspacesParams): Loadable<Workspace[]> => {
   const context = useContext(WorkspacesContext);
   if (context === null) {
     throw new Error('Attempted to use useWorkspaces outside of Workspace Context');
   }
-  const { workspaces } = context;
+  if (context.workspaces === NotLoaded) {
```
Comment (ClaireNeveu) [inline webui/react/src/stores/workspaces.tsx:71]: Nit but I would just use `Loadable.match` here.

File: webui/react/src/stores/workspaces.tsx [hunk 3]
```diff
@@ -53,15 +60,40 @@ export const useEnsureWorkspacesFetched = (canceler: AbortController): (() => Pr
...
+        default:
+          return false;
+      }
+    }, true),
+  );
+  return Loaded(val);
+};
 
-  return workspaces;
+export const updateWorkspace = (id: number, updater: (arg0: Workspace) => Workspace): void => {
```
Comment (ClaireNeveu) [inline webui/react/src/stores/workspaces.tsx:93]: This isn't very useful as is, you'll want to make this like the `useFetchWorkspaces` and return a function that actually does the updating, that way you can call this from a callback.

File: webui/react/src/stores/workspaces.tsx (hunk 2 above)
Comment (ClaireNeveu) [reply webui/react/src/stores/workspaces.tsx:71]: Or just `Loadable.map` since the `NotLoaded` case is just `NotLoaded`.

Comment (mapmeld) [reply webui/react/src/stores/workspaces.tsx]: ✅

File: webui/react/src/stores/workspaces.tsx [hunk 4]
```diff
@@ -53,15 +54,55 @@ export const useEnsureWorkspacesFetched = (canceler: AbortController): (() => Pr
...
+          default:
+            return false;
+        }
+      }, true),
+    ),
+  );
+};
+
+export const useUpdateWorkspace = (
+  canceler: AbortController,
```
Comment (ClaireNeveu) [inline webui/react/src/stores/workspaces.tsx:86]: This argument is unused.

**Pull request merged**